internal/kubernetes/token/grpc.go

```
@@ -0,0 +1,162 @@
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"math/big"
+	"os"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+const bearerPrefix = "Bearer "
+
+func WithServiceAccountToken() grpc.DialOption {
+	return grpc.WithUnaryInterceptor(addAuthorizationHeader)
+}
+
+func addAuthorizationHeader(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	token, err := getToken()
+	if err != nil {
+		return err
+	}
+
+	authCtx := metadata.AppendToOutgoingContext(ctx, "Authorization", "Bearer "+token)
+	return invoker(authCtx, method, req, reply, cc, opts...)
+}
+
+func getToken() (string, error) {
+	return readFile("/var/run/secrets/kubernetes.io/serviceaccount/token")
+}
+
+func AuthorizationInterceptor(kubeclient kubernetes.Clientset) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if err := authorizeConnection(ctx, kubeclient); err != nil {
+			return nil, err
+		}
+		return handler(ctx, req)
+	}
+}
+
+func authorizeConnection(ctx context.Context, kubeclient kubernetes.Clientset) error {
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	klog.Infof("Incoming metadata: %v", md)
+	if !ok {
+		return status.Error(codes.Unauthenticated, "missing metadata")
+	}
+
+	authHeader, ok := md["authorization"]
+	if !ok || len(authHeader) == 0 {
+		return status.Error(codes.Unauthenticated, "missing authorization token")
+	}
+
+	token := authHeader[0]
+	isValidated, err := validateBearerToken(ctx, token, kubeclient)
+	if !isValidated || (err != nil) {
+		return status.Errorf(codes.Unauthenticated, "invalid token")
+	}
```

err will be lost with current logic.
split it into two separate if checks, one for isValidated and one for err